routers/user/auth.go

```diff
@@ -582,28 +582,52 @@ func SignInOAuthCallback(ctx *context.Context) {
 
 	u, gothUser, err := oAuth2UserLoginCallback(loginSource, ctx.Req.Request, ctx.Resp)
 
-	handleOAuth2SignIn(u, gothUser, ctx, err)
-}
-
-func handleOAuth2SignIn(u *models.User, gothUser goth.User, ctx *context.Context, err error) {
 	if err != nil {
 		ctx.ServerError("UserSignIn", err)
 		return
 	}
 
 	if u == nil {
-		// no existing user is found, request attach or new account
-		err = ctx.Session.Set("linkAccountGothUser", gothUser)
-		if err != nil {
-			log.Error(fmt.Sprintf("Error setting session: %v", err))
+		if setting.OAuth2Client.EnableAutoRegistration {
+			// create new user with details from oauth2 provider
+			var name string
+			if setting.OAuth2Client.UseNickname {
+				name = gothUser.NickName
+			} else {
+				name = gothUser.UserID
+			}
+			u = &models.User{
+				Name:        name,
+				FullName:    gothUser.Name,
+				Email:       gothUser.Email,
+				IsActive:    !setting.OAuth2Client.RegisterEmailConfirm,
+				LoginType:   models.LoginOAuth2,
+				LoginSource: loginSource.ID,
+				LoginName:   gothUser.UserID,
```

Comment:
@lunny one thing I didnt looked ad: does this way also migrate migrated-contend to the user?